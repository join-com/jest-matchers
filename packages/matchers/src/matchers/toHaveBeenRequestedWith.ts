import { diff, matcherHint, printExpected, printReceived, RECEIVED_COLOR } from 'jest-matcher-utils'

type ICallWithRequest = [{ request: Record<string, unknown> }]

interface IMock {
  mock: {
    calls: ICallWithRequest[]
  }
}

export const toHaveBeenRequestedWith = ({ mock }: IMock, expected: unknown): jest.CustomMatcherResult => {
  const calls = mock.calls
  const pass = calls.length > 0 && anyOfRequestsSucceeded(calls, expected)
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toHaveBeenRequestedWith')
      const expectedString = printExpected(expected)
      const mismatchString = formatMismatchedCalls(mock.calls, expected)
      return `${hintString}\n\n` + `Expected request value to equal:\n${expectedString}\n` + mismatchString
    },
  }
}

const formatMismatchedCalls = (calls: ICallWithRequest[], expected: unknown): string => {
  if (!calls.length) {
    return `But it was ${RECEIVED_COLOR('not called')}.`
  }

  const messages: string[] = ['Received:']
  calls.map((call, index) => {
    const received = call[0].request
    const diffString = diff(expected, received) ?? ''

    const receivedString = printReceived(received)
    const callNumber: string = calls.length > 1 ? `\ncall no.${String(index + 1)}\n` : ''
    const callMessage = `${callNumber}` + `${receivedString}\n` + 'Difference:\n' + `${diffString}`

    messages.push(callMessage)
  })

  if (calls.length > 1) {
    messages.push(`\nNumber of calls: ${calls.length}`)
  }
  return messages.join('\n')
}

const anyOfRequestsSucceeded = (calls: ICallWithRequest[], expected: unknown): boolean => {
  for (const call of calls) {
    if (equal(call[0].request, expected)) {
      return true
    }
  }
  return false
}

const equal = (request: Record<string, unknown>, expected: unknown): boolean => {
  try {
    expect(request).toEqual(expected)
    return true
  } catch (e) {
    return false
  }
}
