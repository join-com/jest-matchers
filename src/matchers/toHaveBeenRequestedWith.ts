import { equals } from 'expect/build/jasmine_utils';
import diff from 'jest-diff';
import {
  matcherHint,
  printExpected,
  printReceived,
  RECEIVED_COLOR,
} from 'jest-matcher-utils';

export const toHaveBeenRequestedWith = (
  { mock }: jest.Mock<any>,
  expected: object,
) => {
  const calls = mock.calls;
  const pass = calls.length > 0 && equals(calls[0][0].request, expected);
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toHaveBeenRequestedWith');
      const expectedString = printExpected(expected);
      const mismatchString = formatMismatchedCalls(mock.calls, expected);
      return (
        `${hintString}\n\n` +
        `Expected request value to equal:\n${expectedString}\n` +
        mismatchString
      );
    },
  };
};

const formatMismatchedCalls = (calls, expected): string => {
  if (!calls.length) {
    return `But it was ${RECEIVED_COLOR('not called')}.`;
  }

  const received = calls[0][0].request;
  const diffString = diff(expected, received);
  const receivedString = printReceived(received);
  return `Received:\n${receivedString}\n` + `Difference:\n${diffString}`;
};
