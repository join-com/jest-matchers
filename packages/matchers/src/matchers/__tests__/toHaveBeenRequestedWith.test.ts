import { toHaveBeenRequestedWith } from '../toHaveBeenRequestedWith'

expect.extend({ toHaveBeenRequestedWith })

describe('toHaveBeenRequestedWith', () => {
  const mockFn = jest.fn()

  const request = {
    a: 1,
    b: 2,
  }

  beforeEach(() => {
    mockFn.mockClear()
  })

  it('passes when mock was called with request', () => {
    mockFn({ request })

    expect(mockFn).toHaveBeenRequestedWith({
      a: 1,
      b: 2,
    })
  })

  it('passes with objectContaining', () => {
    mockFn({ request })

    expect(mockFn).toHaveBeenRequestedWith(expect.objectContaining({ a: 1 }))
    expect(mockFn).toHaveBeenRequestedWith(expect.objectContaining({ b: 2 }))
  })

  it('fails when mock was not called', () => {
    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 2,
      })
    }

    expect(validate).toThrowErrorMatchingSnapshot()
  })

  it('fails when mock was called with different values', () => {
    mockFn({ request })

    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 1,
      })
    }

    expect(validate).toThrowErrorMatchingSnapshot()
  })

  it('passes if any calls was correct', () => {
    const falseRequest = {
      a: 1,
      b: 1,
    }
    mockFn({ request: falseRequest })
    mockFn({ request })
    mockFn({ request: falseRequest })

    expect(mockFn).toHaveBeenRequestedWith({
      a: 1,
      b: 2,
    })
  })

  it('fails when all calls were incorrect', () => {
    const falseRequest = {
      a: 1,
      b: 1,
    }
    mockFn({ request: falseRequest })
    mockFn({ request: falseRequest })
    mockFn({ request: falseRequest })

    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 2,
      })
    }

    expect(validate).toThrowErrorMatchingSnapshot()
  })
})
