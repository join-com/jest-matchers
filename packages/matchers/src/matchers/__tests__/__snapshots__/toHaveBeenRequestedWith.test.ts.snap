// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveBeenRequestedWith fails when all calls were incorrect 1`] = `
"expect(received).toHaveBeenRequestedWith(expected)

Expected request value to equal:
{"a": 1, "b": 2}
Received:

call no.1
{"a": 1, "b": 1}
Difference:
- Expected
+ Received

  Object {
    "a": 1,
-   "b": 2,
+   "b": 1,
  }

call no.2
{"a": 1, "b": 1}
Difference:
- Expected
+ Received

  Object {
    "a": 1,
-   "b": 2,
+   "b": 1,
  }

call no.3
{"a": 1, "b": 1}
Difference:
- Expected
+ Received

  Object {
    "a": 1,
-   "b": 2,
+   "b": 1,
  }

Number of calls: 3"
`;

exports[`toHaveBeenRequestedWith fails when mock was called with different values 1`] = `
"expect(received).toHaveBeenRequestedWith(expected)

Expected request value to equal:
{"a": 1, "b": 1}
Received:
{"a": 1, "b": 2}
Difference:
- Expected
+ Received

  Object {
    "a": 1,
-   "b": 1,
+   "b": 2,
  }"
`;

exports[`toHaveBeenRequestedWith fails when mock was not called 1`] = `
"expect(received).toHaveBeenRequestedWith(expected)

Expected request value to equal:
{"a": 1, "b": 2}
But it was not called."
`;
